test/caption-stream.test.js

```diff
@@ -1786,6 +1897,276 @@ QUnit.test('reset works', function() {
 });
 
 
-QUnit.skip('paint-on display mode', function() {
-  QUnit.ok(false, 'not implemented');
+QUnit.test('paint-on mode', function() {
+  var packets, captions;
+  packets = [
+    // RDC, resume direct captioning, begin display
+    { pts: 1000, ccData: 0x1429, type: 0 },
+    // 'hi'
+    { pts: 2000, ccData: characters('hi'), type: 0 },
+    // EDM, erase displayed memory. Finish display
+    { pts: 3000, ccData: 0x142c, type: 0 }
+  ];
+  captions = [];
+
+  cea608Stream.on('data', function(caption) {
+    captions.push(caption);
+  });
+
+  packets.forEach(cea608Stream.push, cea608Stream);
+
+  QUnit.equal(captions.length, 1, 'detected a caption');
+  QUnit.deepEqual(captions[0], {
+    startPts: 1000,
+    endPts: 3000,
+    text: 'hi',
+    stream: 'CC1'
+  }, 'parsed the caption');
+});
+
+QUnit.test('preserves newlines from PACs in paint-on mode', function() {
+  var captions = [];
+  cea608Stream.on('data', function(caption) {
+    captions.push(caption);
+  });
+
+  [
+    // RDC, resume direct captioning
+    { pts: 1000, ccData: 0x1429, type: 0 },
+    { pts: 1000, ccData: 0x1350, type: 0 },
+    { pts: 2000, ccData: 0x5445, type: 0 },
+    { pts: 2000, ccData: 0x5354, type: 0 },
+    { pts: 3000, ccData: 0x1450, type: 0 },
+    { pts: 3000, ccData: 0x5354, type: 0 },
+    { pts: 4000, ccData: 0x5249, type: 0 },
+    { pts: 4000, ccData: 0x4e47, type: 0 },
+    { pts: 5000, ccData: 0x1470, type: 0 },
+    { pts: 5000, ccData: 0x4441, type: 0 },
+    { pts: 6000, ccData: 0x5441, type: 0 },
+    // EDM, erase displayed memory. Finish display
+    { pts: 6000, ccData: 0x142c, type: 0 }
+  ].forEach(cea608Stream.push, cea608Stream);
+
+  QUnit.equal(captions.length, 1, 'caption emitted');
+  QUnit.equal(captions[0].text, 'TEST\n\nSTRING\nDATA', 'Position PACs were converted to newlines');
+});
+
+QUnit.test('backspaces are reflected in the generated captions (paint-on)', function() {
+  var captions = [];
+  cea608Stream.on('data', function(caption) {
+    captions.push(caption);
+  });
+
+  [ // RDC, resume direct captioning
+    { ccData: 0x1429, type: 0 },
+    // '01'
+    { pts: 0 * 1000, ccData: characters('01'), type: 0 },
+    // backspace
+    { pts: 0 * 1000, ccData: 0x1421, type: 0 },
+    { pts: 1 * 1000, ccData: characters('23'), type: 0 },
+    { pts: 2 * 1000, ccData: 0x1370, type: 0 },
+    { pts: 2 * 1000, ccData: characters('32'), type: 0 },
+    // backspace
+    { pts: 3 * 1000, ccData: 0x1421, type: 0 },
+    { pts: 4 * 1000, ccData: characters('10'), type: 0 },
+    // EDM, erase displayed memory
+    { pts: 5 * 1000, ccData: 0x142c, type: 0 }
+  ].forEach(cea608Stream.push, cea608Stream);
+
+  QUnit.equal(captions.length, 1, 'detected a caption');
+  QUnit.equal(captions[0].text, '310\n\n023', 'applied the backspaces');
```

Comment:
Why does the order of characters in the output appear opposite of what was fed into the stream?

Reply:
It looks confusing because it starts at the base row (row 15). Then `0x1370` moves the cursor up to row 13 before continuing writing.

I added a couple comments to make it slightly more clear.